### PR TITLE
VB-2711 Implement non-association for SQS - storage

### PR DESCRIPTION
**Pull request by AledWEvans**
Vb 2711 implement non association for sqs   storage

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt [hunk 1]
```diff
@@ -39,10 +41,23 @@ class VisitNotificationService(
     }
   }
 
+  private fun isNotADeleteEvent(nonAssociationChangedNotification: NonAssociationChangedNotificationDto): Boolean {
+    val nonAssociations = prisonerService.getOffenderNonAssociationList(nonAssociationChangedNotification.prisonerNumber)
+    return nonAssociations.any { it.offenderNonAssociation.offenderNo == nonAssociationChangedNotification.nonAssociationPrisonerNumber }
+  }
```
Comment (AledWEvans) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt:47]: Should I also check the dates, just dont know if they will exisit on a delete event,.. only one way to find out here,... also I dont know if there will be a race condition between this event updating the prison service and doing the update,... here... only one way to find out

Comment (rj-adams) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt:47]: Visit notification service should be listening for the booking create/update/delete events which have been generated by the visit scheduler - then sending out notifications accordingly.

Visit notifications should be an independent service - not part of the visit scheduler!

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitNotificationEventRepository.kt [hunk 2]
```diff
@@ -1,8 +1,23 @@
...
+import uk.gov.justice.digital.hmpps.visitscheduler.service.NotificationEventType
 
 @Repository
-interface VisitNotificationEventRepository : JpaRepository<VisitNotificationEvent, Int>
+interface VisitNotificationEventRepository : JpaRepository<VisitNotificationEvent, Int> {
+
+  @Query(
+    "SELECT count(ve) > 0" +
+      " FROM visit_notification_event ve " +
+      " WHERE ve.create_timestamp BETWEEN NOW() - INTERVAL '1 MINUTE' AND NOW() " +
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitNotificationEventRepository.kt:15]: what if we remove this 1 minute check altogether even if they are not recent duplicates?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt [hunk 3]
```diff
@@ -39,10 +41,23 @@ class VisitNotificationService(
     }
   }
 
+  private fun isNotADeleteEvent(nonAssociationChangedNotification: NonAssociationChangedNotificationDto): Boolean {
+    val nonAssociations = prisonerService.getOffenderNonAssociationList(nonAssociationChangedNotification.prisonerNumber)
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt:45]: what will happen if this call fails for timeout?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitNotificationEventRepository.kt [hunk 4]
```diff
@@ -1,8 +1,23 @@
...
+      " AND ve.visit_id=:visitId AND ve.type=:#{#nonAssociationEvent.name()}",
+    nativeQuery = true,
+  )
+  fun isEventARecentDuplicate(
+    visitId: Long,
+    nonAssociationEvent: NotificationEventType,
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitNotificationEventRepository.kt:21]: should var name be changed from nonAssociationEvent to notificationEvent?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt (hunk 1 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt:47]: good point. I think the getOffenderNonAssociationList will return the current non associations while the event could be for a future date. We will have to check the API docs.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitNotificationEventRepository.kt (hunk 2 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/VisitNotificationEventRepository.kt:15]: maybe this is needed as we can have a non association for a 3rd prisoner added.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt (hunk 1 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt:56]: added extra checks

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt (hunk 3 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitNotificationService.kt:54]: good point

**Review by dhirajshettymoj — APPROVED** [review]
LGTM

**Pull request merged**